### ISSUE TITLE
Use consistent Travis configuration files across gems and documents

### ISSUE THREAD

**Issue by ronaldtse** [metanorma/metanorma#48]
Originally from https://github.com/metanorma/unicode2latex/pull/3

This issue is to apply the same Travis configuration across:

* Metanorma gems
* Metanorma gems (with atypical dependencies, e.g. external tools)
* Metanorma documents

---

Original discussion:

> But I think we have more then 2 types, because some of gem required xml2rfx, pupeeteer, plantuml

Yes. Those are the outliers that we have to handle separately.

> 'plain' gem
> gem with external tools
> individual docs

Agree.

> Looks like no way https://stackoverflow.com/questions/44108837/sharing-travis-configuration-among-multiple-repositories but let me check official documentation

Right, but we can use a common `.travis.yml` from a common place (e.g. `metanorma-build-scripts`).

Can you help proceed? Thanks!

**Comment by CAMOBAP795** [comment]
> Right, but we can use a common .travis.yml from a common place (e.g. metanorma-build-scripts).

I'm not sure about this, but https://github.com/travis-ci/travis.rb can help us to manage configuration

Anyway let me first finish #46 I will see whole picture once it will be done

**Comment by CAMOBAP795** [comment]
### Research

I have finished the research and looks like there is no existing solution for this:
 - no way to keep `.travis.yml` config
 - no way to use `travis.rb` to modify all repositories at once (master script need to be written for this)

### Proposed solution

The only solution which I see right now is to create configuration based on `repo` tool https://gerrit.googlesource.com/git-repo/ :
0. Use https://github.com/metanorma/metanorma-build-scripts as repo to keep configuration
1. This repo will contains master copies for `travis.yml` and `appveyour.yml` configurations
1. Create Manifest with all repos which are used CI
2. Create configuration (simple yml, json or xml) called `ci.yml`. which keep a relation between repository and master `travis.yml` and `appveyour.yml` config file for it from https://github.com/metanorma/metanorma-build-scripts
3. Write simple script `sync.rb` (on bash/ruby/python) which will perform 'synchronization' (or propogation) of travis/appveyor config from https://github.com/metanorma/metanorma-build-scripts to target repository
4. This script will create PR

#### Possible`ci.yml` config format
```yml
repos:
  - metanorma/metanorma-cli
    - appveyor.yml: appveyor/appveyor.v2.yml
    - .travis.yml: travis/travis.v5.yml
  - metanorma/metanorma
     - appveyor.yml: appveyor/appveyor.v1.yml
     - .travis.yml: travis/travis.v0.yml
  - ... # another repos
settings:
  force-push: false # push without PR
  ... # anothe settings
```

### Workflow (possible for different use cases)

#### Case 1: Update travis.yml for all repos

1. Update all `travis.yml`s in https://github.com/metanorma/metanorma-build-scripts
2. Run script to create PR for each repo which will be modified

#### Case 2: Modif travis.yml for single repo

1. Create separate  `travis.yml` or `appveyour.yml`
2. Update `ci.yml` to point to this new file
3. Run script to create PR for each repo which will be modified

#### Case 3: Remove from CI

1. Delete repository entry from `ci.yml`
2. Run script to create PR for each repo which will be modified

**Comment by CAMOBAP795** [comment]
@ronaldtse @opoudjis how do you think about comment above?

**Comment by opoudjis** [comment]
> Write simple script sync.rb (on bash/ruby/python) which will perform 'synchronization' (or propogation) of travis/appveyor config from https://github.com/metanorma/metanorma-build-scripts to target repository
> This script will create PR

If this is a one off operation, then sure, what do I care.

If I have to run the script every time the central Travis repo changes, well, I can get around to it, or you can do it. Not that big a deal.

... If I have to manually remember to do this every time I do a release, I will not do it. I'd rather the Travis instances stay redundant.

**Comment by CAMOBAP795** [comment]
> ... If I have to manually remember to do this every time I do a release, I will not do it. I'd rather the Travis instances stay redundant.

This script will not correlate with releases, it just to keep CI (Travis/Appveyor) configuration consistent across many repositories, and will be executed on-demand 

Let me show an example when we may use this script: we need to add a new ruby version to our CI 2.6.x and remove 2.3.x

Currently, I need to create every PR manually, but with this script, I will need to:
 - Update 'master' configuration for travis/appveyor in https://github.com/metanorma/metanorma-build-scripts
 - Run script to update repositories
 - The script will create PRs for all repositories which contain changes

@opoudjis how do you think?

**Comment by opoudjis** [comment]
A-OK, as long as it's not routine.

**Comment by ronaldtse** [comment]
@CAMOBAP795 thanks, I think this approach is good to go. Thanks!

**Comment by ronaldtse** [comment]
(Please write it in Ruby using the GitHub gem, thanks!)

**Comment by ronaldtse** [comment]
@CAMOBAP795 is this being done? Thanks!

**Comment by CAMOBAP795** [comment]
@ronaldtse I would say %70 done, here is intermediate results https://github.com/metanorma/metanorma-build-scripts/tree/feature/ci-master-tool

I plan to finish it in the next 2 days

**Comment by opoudjis** [comment]
Btw, whatever happens, can we start dropping support for Ruby 2.3? We need to reduce execution time drastically on regression testing.

**Comment by ronaldtse** [comment]
@opoudjis yes, we can. We will need to make sure that those people that run Ruby 2.3 gets an error message - gemspec (https://stackoverflow.com/questions/5004492/how-do-you-specify-a-minimum-ruby-version-in-a-gemspec)

**Comment by opoudjis** [comment]
So while I am presumably not touching any of the Travis configs to drop Ruby 2.3, because that's this ticket, once it is in place, I have a separate task to break Ruby 2.3 on all gemspecs.

**Comment by opoudjis** [comment]
OK, I think we can close this, thank you.

**Comment by opoudjis** [comment]
Not so much:

`repo` does not work for me: From what I gather, I need to install git-repo, which you have not said in the README. Please refer README to https://source.android.com/setup/build/downloading

**Comment by CAMOBAP795** [comment]
@opoudjis new PR to improve README.md added https://github.com/metanorma/metanorma-build-scripts/pull/8

> repo sync issue

Probably related to the fact that PR wasn't merged to master at that time, could you please confirm that issue still exists?

**Comment by CAMOBAP795** [comment]
> I suspect your config presupposes familiarity with Python (and that repo was already installed). It should not.

@opoudjis actually there is no need any diving into python

Generally, the solution consists of several tools (which implemented with different technologies):
 - [repo](https://source.android.com/setup/develop/repo) - allows to clone set of repos in 2 commands, support both python2 & python3
 - [git-multi](https://github.com/tkrajina/git-plus) - allows to execute `git COMMAND` in a batch mode, support python2 only
 - [hub](https://github.com/github/hub) - allows create PR from terminal, implemented in golang
 - [ci-master](https://github.com/metanorma/metanorma-build-scripts/tree/master/ci-master) - ruby

@opoudjis to fix git-multi python issue could you please locally change `#!/usr/bin/python` to `#!/usr/bin/python2`

Let me know if you still have some concern about this

**Comment by opoudjis** [comment]
>> repo sync issue
> Probably related to the fact that PR wasn't merged to master at that time, could you please confirm that issue still exists?

Issue still exists, and since it's a permissions issue, and the only update has been in the readme, I don't see why it wouldn't still exist.

````
Nick-Nicholas-Macbook-Pro-Retina:mn-root nickn$ repo sync
Fetching project metanorma/relaton-bib
Fetching project metanorma/metanorma-model-ogc
Fetching project metanorma/ogc-tcpnp
Fetching project metanorma/metanorma-csd
git@github.com: Permission denied (publickey).
git@github.com: Permission denied (publickey).
git@github.com: Permission denied (publickey).
fatal: Could not read from remote repository.

Please make sure you have the correct access rights
and the repository exists.
fatal: Could not read from remote repository.

Please make sure you have the correct access rights
and the repository exists.
git@github.com: Permission denied (publickey).
fatal: Could not read from remote repository.

Please make sure you have the correct access rights
and the repository exists.
git@github.com: Permission denied (publickey).
fatal: Could not read from remote repository.

Please make sure you have the correct access rights
and the repository exists.
git@github.com: Permission denied (publickey).
fatal: Could not read from remote repository.

Please make sure you have the correct access rights
and the repository exists.
git@github.com: Permission denied (publickey).
fatal: Could not read from remote repository.

Please make sure you have the correct access rights
and the repository exists.
git@github.com: Permission denied (publickey).
fatal: Could not read from remote repository.

Please make sure you have the correct access rights
and the repository exists.
git@github.com: Permission denied (publickey).
fatal: Could not read from remote repository.

Please make sure you have the correct access rights
and the repository exists.
error: Cannot fetch metanorma/metanorma-csd from ssh://git@github.com/metanorma/metanorma-csd
Fetching project metanorma/chocolatey-latexml
error: Cannot fetch metanorma/relaton-bib from ssh://git@github.com/metanorma/relaton-bib
error: Cannot fetch metanorma/ogc-tcpnp from ssh://git@github.com/metanorma/ogc-tcpnp
git@github.com: Permission denied (publickey).
fatal: Could not read from remote repository.

Please make sure you have the correct access rights
and the repository exists.
error: Cannot fetch metanorma/metanorma-model-ogc from ssh://git@github.com/metanorma/metanorma-model-ogc
git@github.com: Permission denied (publickey).
fatal: Could not read from remote repository.

Please make sure you have the correct access rights
and the repository exists.
error: Cannot fetch metanorma/chocolatey-latexml from ssh://git@github.com/metanorma/chocolatey-latexml

error: Exited sync due to fetch errors
````

**Comment by opoudjis** [comment]
> git -C ../../ multi checkout master

That's bad practice, hardcoding a relative path between two independent folders; you did not say at the start that mn-root is a sister directory of metanorma-build-scripts, and you shouldn't be assuming it is. You should be saying `git -C {mn-root} multi checkout master`, and let the user deal with populating where they have created mn-root.

**Comment by ronaldtse** [comment]
@opoudjis is your SSH public key uploaded to your GitHub account for access?

**Comment by ronaldtse** [comment]
And indeed hardcoding relative paths is bad. That’s why I said should extract ci-master.

**Comment by opoudjis** [comment]
> @opoudjis is your SSH public key uploaded to your GitHub account for access?

No. I take it it is supposed to be?

**Comment by ronaldtse** [comment]
To be able to access GitHub repos via SSH your SSH public key needs to be associated with your GitHub account.

**Comment by opoudjis** [comment]
Error in repo sync persists after me giving github my ssh. I've had to enter my SSH password several dozen times, once apparently for each repository, and it only ended up checking out two repositories.

Getting this working is a very low priority for me.

**Comment by ronaldtse** [comment]
Ah. To get rid of the SSH prompt, you should either be using `ssh-agent`, or do not set a password for the SSH key...

**Comment by opoudjis** [comment]
It worked.

**Comment by opoudjis** [comment]
Worked. There were things that should have been documented better: run bundle install in ci-master, make sure you've got ssh-agent running and an uploaded SSH key, make sure you install git-plus only in Python2 and not Python3 (because it will attempt to run the git-multi as Python3 but it is in fact Python2 code.)

**Comment by CAMOBAP795** [comment]
> That's bad practice, hardcoding a relative path between two independent folders

They are not, `metanorma-build-scripts` declared in https://github.com/metanorma/metanorma-build-scripts/blob/master/default.xml#L106

So I can not agree that they are independent

> Error in repo sync persists after me giving github my ssh.

@opoudjis I just wondering previously you have used git only via http?

> There were things that should have been documented better

 - "run bundle install in ci-master" - good point, I will add it
 - "make sure you've got ssh-agent running and an uploaded SSH key" - aren't we assume that user should have a basic understanding of how git/ruby works because if don't, we should explicitly note that git/ruby needs to be installed too
 - "make sure you install git-plus only in Python2" - unfortunately sometimes it happens I will do pr for it https://github.com/tkrajina/git-plus/issues/15

cc @ronaldtse

**Comment by opoudjis** [comment]
mn-root is not mentioned in default.xml

I had not set my SSH in my GitHub account settings, so I guess I was only using http. That's certainly all I ever cloned with.

**Comment by CAMOBAP795** [comment]
> mn-root is not mentioned in default.xml

Yes because this isn't a repository, mn-root just a dir to aggregate all repos described in default.xml
